### ISSUE TITLE
Support package specific compiler flag

### ISSUE THREAD

**Issue by Char-Aznable** [LLNL/spack#19052]
### Rationale
Currently, compiler flags are passed down the dependencies. This may not always be desirable and may result in inflating package installation. More often than not, I only want to use `-g` debug flag for package `A` and I don't want the debug flag to be used for every single dependency package. 

### Description

I'd like to have something like:
```
packages:
   trilinos:
    flags:
      cppflags: "-g"
```

### General information

- [x] I have run `spack --version` and reported the version of Spack
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by haampie** [comment]
I totally agree with the problem pointed out in this issue. Additionally there is the issue you cannot specify different compiler flags for different packages on the command line.

Alternative solution: make a breaking change to not pass compiler flags specified on the command line down to dependencies. I believe this is the intuitive behavior.

**Comment by mwkrentel** [comment]
Trilinos is a cmake package, so why doesn't `build_type` do what you
want here?  You can use `spack install trilinos build_type=Debug`
to set C/C++ flags to `-g`.  Or, Release, RelWithDebInfo, MinSizeRel
for other values.

Although I think this is a good idea in general, assuming it's not too
hard to implement, I suggest that treating debug and optimization
flags separately is a better idea.

CMakePackage already does this with `build_type`.  I suggested adding
similar functionality to AutotoolsPackage in #14430.  I'm just waiting
for traction ....

**Comment by Char-Aznable** [comment]
cmake_build_type and potentially similar config option from other package builders are subject to upstream customization, i.e., setting `build_type=Debug` or Release or whatever might not always get you what you want. Treating debug and optimization flags is just a very special case of the OP's proposal
